README
NAME
    CGI::Application::Search - Base class for CGI::App Swish-e site engines

SYNOPSIS
        package My::Search;
        use base 'CGI::Application::Search';

        sub cgiapp_init {
            my $self = shift;
            $self->param(
                'SWISHE_INDEX' => 'my-swishe.index',
                'TEMPLATE'     => 'search_results.tmpl',
            );
        }

        sub cgiapp_prerun {
            my $self  = shift;
            my $query = $self->query;

            # let the user turn context highlighting off
            $self->param('HIGHLIGHT' => 0) if $query->param('highlight_off');

            # let the user change which property is used in the sort
            my $sort_by = $query->param('sort_by');
            $self->param('SORT_BY' => $sort_by) if $sort_by;
        }

        1;

DESCRIPTION
    A CGI::Application based control module that uses Swish-e API in perl
    (<http://swish-e.org>) to to perform searches on a swish-e index of
    documents.

  Features
    * Sub-Classable. Unlike the Perl examples that come with swish-e, this
    is not a script, and can be customized without modifiying the original
    so that several sites may share the same underlying code.
    * Uses CGI::Application::Plugin::AnyTemplate to allow flexibility in
    template engine choice (HTML::Template, Template-Toolkit or Petal).
    * Built-in templates to use out of box or as examples for your own
    templates
    * HiLighted search results
    * HiLighted pages linked from search results
    * AJAX results sent to page without need of a page reload
    * AJAX powered 'auto-suggest' to give the user list of choices available
    for search

TUTORIAL
    If this is your first time using Swish-e (or you think you need a
    refresher) or if you want step-by-step instructions on how to use the
    AJAX capabilities of this module, then please see the "Tutorial".

RUN_MODES
    The start_mode is show_search.

  show_search()
    This method will load the template pointed to by the TEMPLATE param
    (falling back to a default internal template if none is specified) and
    display it to the user. It will 'associate' this template with $self so
    that any parameters in $self->param() are also accessible to the
    template. It will also use HTML::FillInForm to fill in the search form
    with the previously selected parameters (unless it's a
    'non-page-refresh' AJAX search).

  perform_search()
    This is where the meat of the searching is performed. We create a
    SWISH::API object on the SWISHE_INDEX and create the query for the
    search based on the value of the *keywords* parameter in CGI and any
    other EXTRA_PARAMETERS. The search is executed and if HIGHLIGHT is true
    we will use Search::Tools::HiLiter to highlight it and then format the
    results, only showing PER_PAGE number of elements A paging list is also
    shown for navigating through the results. Then we will return to the
    show_search() method for displaying everything.

  highlight_remote_page
    This run mode will fetch a remote page (with either a relative, or
    absolute URL using the "url" Query param) and highlight the keywords
    used in the search on that page (the "keywords" Query param) using the
    HIGHLIGHT_TAG, HIGHLIGHT_CLASS or HIGHLIGHT_COLORS options. This run
    mode is best used in the links of the search results listing.

        <a href="?rm=highlight_remote_page;url=http%3A%2F%2Fexample.com%2Fabout_us%2Findex.html;keywords=Us">about us</a>

  highlight_local_page
    This run mode will fetch a local page (only allowing relative files
    based in the DOCUMENT_ROOT config var and the path using the "path"
    Query param) and highlight the keywords used in the search on that page
    (the "keywords" Query param) using the HIGHLIGHT_TAG, HIGHLIGHT_CLASS or
    HIGHLIGHT_COLORS options. This run mode is best used in the links of the
    search results listing.

        <a href="?rm=highlight_local_page;path=%2Fabout_us%2Findex.html;keywords=Us">about us</a>

  suggestions
    This run mode will return an AJAX listing of words that should be
    suggested to the user for the words that they have typed so far. It uses
    the "suggested_words()" method to actually choose which words to send
    back.

OTHER METHODS
    Most of the time you will not need to call the methods that are
    implemented in this module. But in some cases customizing the templates
    is not enought. If so, it might be prudent to override or extend these
    methods in your derived class.

  new()
    We simply override and extend the CGI::Application new() to setup our
    defaults.

  setup()
    Here's were we setup our run modes. If you override this method, make
    sure you also call it in your base class

        sub setup {
            my $self = shift;
            # do your thing
            ...
            $self->SUPER::setup();
        }

  generate_search_query($keywords)
    This method is used to generate the query for swish-e from the $keywords
    (by default the 'keywords' CGI parameter), as well as any
    EXTRA_PROPERTIES that are present.

    If you wish to generate your own search query then you should override
    this method. This is common if you need to have access/authorization
    control that will need to be taken into account for your search. (eg,
    anything under /protected can't be seen by someone not logged in).

    Please see the swish-e documentation on the exact syntax for the query.

  suggested_words($word)
    This object method is used by the AUTO_SUGGEST flag to return the words
    that should be suggested to the user after they have typed a $word. It
    returns an array reference of those words.

    By default it will just look for words in the AUTO_SUGGEST_FILE that
    begin with $word. If you need more performance or flexibility (eg,
    storing your words in a database and querying for them) you are
    encouraged to override this method.

CONFIGURATION
    There are several configuration parameters that you can set at any time
    (using "param()" in your cgiapp_init, or PARAMS hash in new()) before
    the run mode is called that will affect the search and display of the
    results. They are:

  SWISHE_INDEX
    This is the swishe index used for the searches. The default is
    'data/swish-e.index'. You will probably set this every time.

  AJAX
    This is a boolean indicating whether or not a non-page-refresh AJAX
    search will be permitted.

    Please see the "Tutorial" for more information on how to use the AJAX
    capabilities of this module.

  TEMPLATE
    The name of the search interface template. Default templates are
    included with this distribution and will be used if you don't specify
    one. Which default template is used depends on which TEMPLATE_TYPE you
    are using (*HTMLTemplate* or *TemplateToolkit*) and whether or not the
    AJAX flag is true.

    These sample templates are installed with the module, or you can view
    them by looking under the templates/ directory of the source
    distribution (*.tar.gz*).

    Please see "TEMPLATE USAGE" for more information on which variables are
    passed into your template.

  TEMPLATE_TYPE
    This module uses CGI::Application::Plugin::AnyTemplate to allow
    flexibility in choosing which templating system to use for your search.
    This works especially well when you are trying to integrate the Search
    into an existing app with an existing templating structure.

    This value is passed to the "$self->template->config()" method as the
    "default_type". By default it is 'HTMLTemplate'. Please see
    CGI::Application::Plugin::AnyTemplate for more options.

    If you want more control of configuration for the template the it would
    probably best be done by subclassing CGI::Application::Search and
    passing your desired params to "$self->template->config".

  PER_PAGE
    How many search result items to display per page. The default is 10.

  HIGHLIGHT
    Boolean indicating whether or not we should highlight the description
    given to the templates. The default is true.

  HIGHLIGHT_TAG
    The HTML tag used to surround the highlighted context. The default is
    "strong".

  HIGHLIGHT_CLASS
    The class attribute of the HIGHLIGHT_TAG HTML tag. This is useful when
    you want to dictacte the style through a CSS style sheet. If given, this
    value will override that of HIGHLIGHT_COLORS. It has no value by
    default.

  HIGHLIGHT_COLORS
    This is an array ref of acceptable HTML colors. If provided, it will
    highlight each matching word/phrase in an alternating style. For
    instance, if given 2 colors, every other highlighted phrase would be a
    different color. By default it is an empty array.

  EXTRA_PROPERTIES
    This is an array ref of extra properties used in the search. By default,
    the module will only use the value of the 'keywords' parameter coming in
    the CGI query. If anything is provided as an extra property then it will
    be added to the query used in the search.

    An example: You have some of you pages designated into categories. You
    want the user to have the option of narrowing his results by category.
    You add the word 'category' to the 'EXTRA_PROPERTIES' list and then you
    add a 'category' form element that the user has the option of giving a
    value to your search form. If the user gives that element a value, then
    it will be seen and applied to the search. This will also only work if
    you have the 'category' element defined for your documents (see *SWISH-E
    Configuration* and 'MetaNames' in the swish-e.org SWISH-CONF
    documentation).

    By default, this list is empty.

  EXTRA_RANGE_PROPERTIES
    This is almost exactly like the "EXTRA_PROPERTIES" above except that
    instead of searching for the given properties as simple strings, we will
    use a range. Since ranges need to values, if you're searching for the
    "foo" property, then you need to have a "foo_start" and a "foo_end"
    value coming from the query. So if "foo" is in your
    "EXTRA_RANGE_PROPERTIES" and you have a CGI query string like this:

        ?foo_start=123&foo_end=234

    Then we will a generate a Swish-E query that looks something like this:

      -L foo 123 234

  DESCRIPTION_LENGTH
    This is the maximum length for the context (in chars) that is displayed
    for each search result. The default is 250 characters.

  DESCRIPTION_CONTEXT
    This is the number of words on either side of the searched for words and
    phrases (keywords) that will be displayed as part of the description. If
    this is 0, then the entire description will be displayed. The default is
    0.

    NOTE: This directive will cause Search to use Text::Context, which can
    be slow and CPU intensive at times. These computations may prove to be
    too much for some servers (eg, a shared hosting environment).

  AUTO_SUGGEST
    If true, then this will allow the broswer to give suggestions to the
    user as they type. To use this, you must either use the
    AUTO_SUGGEST_FILE configuration option, or override the
    "suggested_words()" method.

    You template must also have the appropriate JavaScript code. Please see
    the "Tutorial" for more details.

  AUTO_SUGGEST_FILE
    The name of the file where the suggested words are stored. These words
    should be in alphabetical order with one word per line.

  AUTO_SUGGEST_CACHE
    A boolean indicating whether or not the results of the AUTO_SUGGEST_FILE
    should be cached in memory or not. This will save repeated file accesses
    when used in a persistant environment.

  AUTO_SUGGEST_LIMIT
    An integer count of the most suggestions to show the user at a time.
    This is useful when you don't want to overwhelm the end user and take
    over their screen with all of your helpful suggestions.

  DOCUMENT_ROOT
    This is the root directory to use when looking for files when using the
    "highlight_local_page" run mode.

  SORT_BY
    This is a string used by Swish-e to sort the results. The string is a
    space separated list of valid document properties. Each property may
    contain a qualifier (either "asc" or "desc") that sets the direction of
    the sort. Leave it alone and Swish-e will sort by "swishrank" in
    descending order. But say you wanted to reverse that for some reason.
    You could specify a "SORT_BY" of

        swishrank asc

  FIF_DEFAULTS
    A hash reference of extra parameters that are passed to HTML::FillInForm
    when it does it's business with the forms. This is useful to telling
    HTML::FillInForm to skip certain forms or form elements.

TEMPLATE USAGE
    Sample templates are included with this distribution. These sample
    templates are installed with the module, or you can view them by looking
    under the templates/ directory of the source distribution (*.tar.gz*).

    Please feel free to copy and change them in what ever way you see fit.
    To help give you more information to display (or not display, depending
    on your preference) the following variables are available for your
    templates:

  Global Tmpl Vars
    These variables are available throughout the templates and contain
    information related to the search as a whole:

    * ajax  A boolean indicating whether or not this search is an AJAX
            search or not. You can use this flag to exclude everything but
            your search results in your template.

    * url   The URL of this application. This is useful if you want to use
            the same templates in multiple applications, especially if you
            are using the AJAX capabilities since they require the URL to
            submit to.

    * searched
            A boolean indicating whether or not a search was performed.

    * keywords
            The exact string that was recieved by the server from the input
            named 'keywords'

    * elapsed_time
            A string representing the number of seconds that the search
            took. This will be a floating point number with a precision of
            3.

    * hits  This is an array of hashs (TMPL_LOOP in H::T) that contains one
            entry for each result returned (for the current page). Each
            entry contains the following keys:

            reccount
                    The "swishreccount" property of the results as indexed
                    by SWISH-E

            rank    The rank to the result as given by SWISH-E (the
                    "swishrank" property)

            title   The "swishtitle" property of the results as indexed by
                    SWISH-E

            path    The "swishdocpath" property of the results as indexed by
                    SWISH-E

            last_modified
                    The "swishlastmodified" property of the results as
                    indexed by SWISH-E and then formatted using
                    Time::Piece's "strftime()" method with a format string
                    of "%B %d, %Y".

            size    The "swishdocsize" property of the results as indexed by
                    SWISH-E and then formatted with Number::Format's
                    "format_bytes()" method.

            description
                    The "swishdescription" property of the results as
                    indexed by SWISH-E. If HIGHLIGHT is true, then this
                    description will also have search terms highlighted and
                    will only be, at most, DESCRIPTION_LENGTH characters
                    long.

    * pages This is an array of hashes (TMPL_LOOP in H::T) that contains
            paging information for the results. It contains the following
            keys:

            current A boolean indicating whether or not this iteration is
                    the current page or not.

            page_num
                    The integer number of the page.

    * first_page
            This is a boolean indicating whether or not this page of the
            results is the first or not.

    * last_page
            This is a boolean indicating whether or not this page of the
            results is the last or not.

    * prev_page
            The integer number of the previous page. Will be 0 if there is
            no previous page.

    * next_page
            The integer number of the next page. Will be 0 if there is no
            next page.

    * start_num
            This is the number of the first result on the current page

    * stop_num
            This is the number of the last result on the current page

    * total_entries
            The total number of results in their search, not the total
            number shown on the page.

OTHER NOTES
    *   If at any time prior to the execution of the 'perform_search' run
        mode you set the "$self->param('results')" parameter, a search will
        not be performed. Instead those results are returned. This is
        helpful when you decide in the "cgiapp_init" stage that this user
        does not have permissions to perform the desired search.

    *   You must use the *StoreDescription* setting in your Swish-e
        configuration file. If you don't you'll get an error when
        C::A::Search tries to retrieve a description for each hit.

AUTHOR
    Michael Peters <mpeters@plusthree.com>

    Thanks to Plus Three, LP (http://www.plusthree.com) for sponsoring my
    work on this module.

CONTRIBUTORS
    Sam Tregar <sam@tregar.com>
    Mark Stosberg <mark@summersault.com>
    Eric Folley <efolley@plusthree.com>



Changes
Revision history for Perl extension CGI::Application::Dispatch.

1.13 (Apr 27, 2010)
    - Fixed broken tests that were using a wrong date pattern [Jason Purdy]

1.12 (Apr 21, 2010)
    - Added FIF_DEFAULTS parameter to control HTML::FillInForm [Brad Oaks]

1.11 (Nov 17, 2009)
    - Better diagnostics when we can't load SWISH::API to be able to track down the reasons [M. Stosberg]
    - If we can't load Search::Tools::HiLiter then just warn and move one without highlighting
    - UTF-8 is now the default.

1.10 (Nov 16, 2009)
    - UTF-8 can now be turned on with UTF8 parameter
    - Using Search::Tools::HiLiter instead of HTML::HiLiter

1.05 (Jan 02, 2007)
    - Can now search on poperties that are false, but still there (I'm thinking of you
      zero)
    - Added SORT_BY configuration param to change the order.

1.04 (Nov 06, 2007)
    - Fixed "associate" regression where data in $self->param wasn't
      getting sent to the templates.
    - Fixed broken Makefile.PL
    - Adding new EXTRA_RANGE_PROPERTIES to the available fields returned in the results

1.03 (Nov 05, 2007)
    - Changed generate_search_query to allow searching when no
      key words are specified. The search can be on other properties
      or for everything listed. NOTE: This breaks backwards compatability
      if you were relying on empty searches returning no results.

1.02 (Nov 05, 2007)
    - Now using swish-e 2.4.5
    - Added EXTRA_RANGE_PROPERTIES

1.00 (Jul 08 2006)
    - Added Test::LongString and Test::Class to build_requires
    - Fixed bug with URL generation not including PATH_INFO
    - Fix bug in Ajax TT template that didn't work in non-ajax browsers

0.10 (Nov 21, 2005)
    - Changes to Makefile.PL to make CPAN installation happier
    - documentation cleanup and clarifications

0.09 (Nov 17, 2005)
    - Default templates now fail more gracefully when the browser doesn't support AJAX
    - no longer need both AJAX and AUTO_SUGGEST to use just the AUTO_SUGGEST functionality
    - Makefile.PL not installs sample templates like Build.PL (this should make CPAN happy)

0.08 (Nov 16, 2005)
    - No code changes, just refactored out the repetitive tests using Test::Class
    - move tests and coverage

0.07 (Nov 13, 2005)
    - No code changes just fixed failing tests when installed via CPAN

0.06 (Nov 12, 2005)
    - added AJAX flag to turn on non-refresh searches
    - added AUTO_SUGGEST, AUTO_SUGGEST_FILE, AUTO_SUGGEST_CACHE and AUTO_SUGGEST_LIMIT
        functionality
    - minor doc fixes
    - fixed mispelling of 'global_var' => 'global_vars' in HTML::Template
        options
    - removed use of callbacks since they didn't really add anything
        and required newer versions of C::A
    - better and more complete sample templates that are installed with
        the application and useable when no others are provided.
    - split tutorial out into a separate file and include some setup
        info for AJAX usage.

0.05 (Jul 10, 2005)
    - renamed 'HIGHLIGHT_CONTEXT' to just 'HIGHLIGHT'
    - removed 'HIGHLIGT_STOP' and 'HIGHLIGHT_START' in favor of
        'HIGHLIGHT_TAG' and 'HIGHLIGHT_CLASS' and 'HIGHLIGHT_COLORS'
    - renamed 'CONTEXT_LENGTH' to 'DESCRIPTION_LENGTH'
    - default highlighting is now done with HTML::HiLiter (much faster)
    - DESCRIPTION_CONTEXT added so that highlighting is now independent of
        finding a matching 'context' inside the description

0.04 (Jul 08, 2005)
    - removed 'hit_' prefix from all vars in in the 'hits' loop for templates
    - almost 100% test coverage (99.6%)
    - minor bug fixes with built-in template
    - use Carp for all errors and warnings

0.03 (Jul 05, 2005)
    - changed from straight H::T to CGI::Application::Plugin::AnyTemplate
        for flexibility
    - better template documentation and minor template fixes
    - more complete test suite, still not 100% coverage, but closer

0.02 (May 23, 2005)
    - find and fix document mistakes - Sam Tregar <sam@tregar.com>
    - added tutorial for getting started with swishe - Sam Tregar
    - added TEMPLATE option to setup() - Sam Tregar

0.01 (May 18, 2005)
    - First released version


t/lib/MyTest/AJAXBase.pm
package MyTest::AJAXBase;
use base 'MyTest::Base';
use Test::More;
use Test::LongString;
use File::Spec::Functions qw(catfile);

my $SUGGEST_FILE = catfile('t', 'conf', 'index_words');

sub B_ajax_search: Test(9) {
    my $self = shift;

    # simple word 'please'
    # no higlighting
    my $cgi = CGI->new({
        rm          => 'perform_search',
        keywords    => 'please',
        ajax        => 1,
    });
    my $app = CGI::Application::Search->new(
        QUERY   => $cgi,
        PARAMS  => {
            %MyTest::Base::BASE_OPTIONS,
            $self->options,
            HIGHLIGHT => 0,
        },
    );
    $output = $app->run();
    lacks_string($output, '<h2>Search Results</h2>');
    lacks_string($output, 'No results');
    like($output, qr/Elapsed Time: \d\.\d{1,3}s/i);
    like($output, qr/>\w+ \d\d?, 200\d - \d+(K|M|G)?</i);
    contains_string($output, 'This is a Test');
    contains_string($output, 'Please Help Me');
    contains_string($output, 'This is a test. This is a only a test. And please do not panic.');
    contains_string($output, 'Would you please help me find this document');
    contains_string($output, 'Results: 1 to 2 of 2');
    
    # set back to empty
    %MyTest::Base::BASE_CGI_PARAMS = ();
}

sub L_suggestions: Test(12) {
    my $self = shift;
    my @p_words = qw(
        panic
        paragraph
        please
        porta
        praesent
        pretium
        proin
    );
    my $cgi = CGI->new({
        rm          => 'suggestions',
        keywords    => 'p',
        ajax        => 1,
    });

    # with no limit
    my $app = CGI::Application::Search->new(
        QUERY   => $cgi,
        PARAMS  => {
            %MyTest::Base::BASE_OPTIONS,
            $self->options,
            AUTO_SUGGEST      => 1,
            AUTO_SUGGEST_FILE => $SUGGEST_FILE,
        },
    );
    $output = $app->run();
    my $full_regex = join('.*', map { quotemeta($_) } @p_words);
    like($output, qr/$full_regex/);

    # now set a limit
    $app = CGI::Application::Search->new(
        QUERY   => $cgi,
        PARAMS  => {
            %MyTest::Base::BASE_OPTIONS,
            $self->options,
            AUTO_SUGGEST       => 1,
            AUTO_SUGGEST_LIMIT => 2,
            AUTO_SUGGEST_FILE  => $SUGGEST_FILE,
        },
    );
    $output = $app->run();
    my $partial_regex = join('.*', map { quotemeta($_) } @p_words[0..1]);
    unlike($output, qr/$full_regex/);
    like($output, qr/$partial_regex/);

    # with no limit cached
    $app = CGI::Application::Search->new(
        QUERY   => $cgi,
        PARAMS  => {
            %MyTest::Base::BASE_OPTIONS,
            $self->options,
            AUTO_SUGGEST       => 1,
            AUTO_SUGGEST_CACHE => 1,
            AUTO_SUGGEST_FILE  => $SUGGEST_FILE,
        },
    );
    $output = $app->run();
    like($output, qr/$full_regex/);

    # now set a limit with cached
    $app = CGI::Application::Search->new(
        QUERY   => $cgi,
        PARAMS  => {
            %MyTest::Base::BASE_OPTIONS,
            $self->options,
            AUTO_SUGGEST       => 1,
            AUTO_SUGGEST_LIMIT => 2,
            AUTO_SUGGEST_CACHE => 1,
            AUTO_SUGGEST_FILE  => $SUGGEST_FILE,
        },
    );
    $output = $app->run();
    unlike($output, qr/$full_regex/);
    like($output, qr/$partial_regex/);

    # multiple words
    $cgi = CGI->new({
        rm          => 'suggestions',
        keywords    => 'stuff p',
        ajax        => 1,
    });
    $app = CGI::Application::Search->new(
        QUERY   => $cgi,
        PARAMS  => {
            %MyTest::Base::BASE_OPTIONS,
            $self->options,
            AUTO_SUGGEST       => 1,
            AUTO_SUGGEST_LIMIT => 2,
            AUTO_SUGGEST_CACHE => 1,
            AUTO_SUGGEST_FILE  => $SUGGEST_FILE,
        },
    );
    $output = $app->run();
    $partial_regex = join('.*', map { quotemeta("stuff $_") } @p_words[0..1]);
    like($output, qr/$partial_regex/);

    # without specifiying a file
    $app = CGI::Application::Search->new(
        QUERY   => $cgi,
        PARAMS  => {
            %MyTest::Base::BASE_OPTIONS,
            $self->options,
            AUTO_SUGGEST       => 1,
            AUTO_SUGGEST_LIMIT => 2,
        },
    );
    $self->_throw_away_stderr();
    $output = $app->run();
    $self->_restore_stderr();
    contains_string($output, '<ul></ul>');
    contains_string($self->_stderr_junk(), 'AUTO_SUGGEST_FILE was not specified!');

    # without a readable file
    $app = CGI::Application::Search->new(
        QUERY   => $cgi,
        PARAMS  => {
            %MyTest::Base::BASE_OPTIONS,
            $self->options,
            AUTO_SUGGEST       => 1,
            AUTO_SUGGEST_LIMIT => 2,
            AUTO_SUGGEST_FILE  => 'foo.txt',
        },
    );
    $self->_throw_away_stderr();
    $output = $app->run();
    $self->_restore_stderr();
    contains_string($output, '<ul></ul>');
    contains_string($self->_stderr_junk(), 'AUTO_SUGGEST_FILE foo.txt is not readable!');

    # without AUTO_SUGGEST
    $app = CGI::Application::Search->new(
        QUERY   => $cgi,
        PARAMS  => {
            %MyTest::Base::BASE_OPTIONS,
            $self->options,
        },
    );
    $self->_throw_away_stderr();
    $output = $app->run();
    $self->_restore_stderr();
    contains_string($self->_stderr_junk(), 'without AUTO_SUGGEST turned on');
}

1;




Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.3603
require 5.006001;

    unless (eval "use Module::Build::Compat 0.02; 1" ) {
      print "This module requires Module::Build to install itself.\n";

      require ExtUtils::MakeMaker;
      my $yn = ExtUtils::MakeMaker::prompt
	('  Install Module::Build now from CPAN?', 'y');

      unless ($yn =~ /^y/i) {
	die " *** Cannot install without Module::Build.  Exiting ...\n";
      }

      require Cwd;
      require File::Spec;
      require CPAN;

      # Save this 'cause CPAN will chdir all over the place.
      my $cwd = Cwd::cwd();

      CPAN::Shell->install('Module::Build::Compat');
      CPAN::Shell->expand("Module", "Module::Build::Compat")->uptodate
	or die "Couldn't install Module::Build, giving up.\n";

      chdir $cwd or die "Cannot chdir() back to $cwd: $!";
    }
    eval "use Module::Build::Compat 0.02; 1" or die $@;
    
    Module::Build::Compat->run_build_pl(args => \@ARGV);
    my $build_script = 'Build';
    $build_script .= '.com' if $^O eq 'VMS';
    exit(0) unless(-e $build_script); # cpantesters convention
    require Module::Build;
    Module::Build::Compat->write_makefile(build_class => 'Module::Build');


Build.PL
use Module::Build;
use lib './t/lib';
Module::Build->new(
    module_name => 'CGI::Application::Search',
    license => 'perl',
    requires => {
        'perl'                                    => '5.6.1',
        'CGI::Application'                        => 4.0,
        'Encode'                                  => 0,
        'Unicode::Normalize'                      => 0,
        'CGI::Application::Plugin::AnyTemplate'   => 0,
        'SWISH::API'                              => 0.04, # in swish-e 2.4.5
        'Search::Tools::HiLiter'                  => 0.12,
        'Data::Page'                              => 1.0,
        'File::Spec::Functions'                   => 0,
        'Number::Format'                          => 0,
        'HTML::FillInForm'                        => 1.04,
        'Time::HiRes'                             => 0,
        'Time::Piece'                             => 0,
        'POSIX'                                   => 0,
        'Text::Context'                           => 3.5,
        'File::Spec'                              => 0,
        'Capture::Tiny'                           => 0.06,
        'Cwd'                                     => 0,
        'File::Slurp'                             => 9999.13,
    },
    build_requires  => {
        'Test::LongString'                        => 0,
        'Test::Class'                             => 0,
    },
    create_readme      => 1,
    pm_files => {
        'lib/CGI/Application/Search.pm'      => 'lib/CGI/Application/Search.pm',
        'templates/search_results.tmpl'      => 'lib/CGI/Application/Search/templates/search_results.tmpl',
        'templates/search_results.tt'        => 'lib/CGI/Application/Search/templates/search_results.tt',
        'templates/ajax_search_results.tmpl' => 'lib/CGI/Application/Search/templates/ajax_search_results.tmpl',
        'templates/ajax_search_results.tt'   => 'lib/CGI/Application/Search/templates/ajax_search_results.tt',
        'templates/prototype_javascript.tmpl'=> 'lib/CGI/Application/Search/templates/prototype_javascript.tmpl',
    },
    create_makefile_pl => 'passthrough',
)->create_build_script();
